packages/connect/src/createScClient/createScClient.ts

```diff
@@ -11,19 +20,293 @@ export interface ScClient {
   addChain: (chainSpec: string) => Promise<ProviderInterface>
 }
 
+type ResponseCallback = (response: string | Error) => void
+
+class Provider implements ProviderInterface {
+  readonly #coder: RpcCoder = new RpcCoder()
+  readonly #getChain: (handler: JsonRpcCallback) => Promise<Chain>
+  readonly #subscriptions: Map<
+    string,
+    [ResponseCallback, { method: string; id: string | number }]
+  > = new Map()
+  readonly #orphanMessages: Map<string, Array<string | Error>> = new Map()
+  readonly #requests: Map<number, ResponseCallback> = new Map()
+  readonly #eventemitter: EventEmitter = new EventEmitter()
+  #chain: Promise<Chain> | null = null
+  #isChainReady: boolean = false
+
+  public constructor(getChain: (handler: JsonRpcCallback) => Promise<Chain>) {
+    this.#getChain = getChain
+  }
+
+  public get hasSubscriptions(): boolean {
+    // you would expect this to be `this.#subscriptions.size > 0`, right?
+    // well, nope... This must always return `true` b/c otherwise Polkadot
+    // thinks that this Provider doesn't support subscriptions.
+    return true
+  }
+
+  public get isConnected(): boolean {
+    return !!this.#chain && this.#isChainReady
+  }
+
+  public clone(): ProviderInterface {
+    throw new Error("clone() is not supported.")
+  }
+
+  async connect(): Promise<void> {
+    if (this.isConnected) throw new Error("Already connected!")
+
+    // it could happen that after emitting `disconnected` due to the fact taht
+    // smoldot is syncing, the consumer tries to reconnect after a certain amount
+    // of time... In which case we want to make sure that we don't create a new
+    // chain.
+    if (this.#chain) {
+      await this.#chain
+      return
+    }
+
+    const hc = healthChecker()
+    const onResponse = (res: string): void => {
+      if (!hc.responsePassThrough(res)) {
+        return
+      }
+
+      const response = JSON.parse(res) as JsonRpcResponse
+
+      // the healthChecker alters the original id. So, we must restore it
+      // because otherwise the RpcCoder won't be able to decode the resonse.
+      if (
+        typeof response.id === "string" &&
+        (response.id as unknown as string).startsWith("extern:")
+      ) {
+        response.id = Number((response.id as unknown as string).slice(7))
+      }
+
+      let decodedResponse: string | Error
+      try {
+        decodedResponse = this.#coder.decodeResponse(response) as string
+      } catch (e) {
+        decodedResponse = e as Error
+      }
+
+      // It's not a subscription message, but rather a standar RPC response
+      if (response.method === undefined) {
```

Comment:
I think you should check this with `response.params.subscription`, not with  ̀response.method`

Reply:
I think it's the same thing, but I don't mind changing it...